llm-finetuning/huggingface/hf_deployment_service.py

```diff
@@ -0,0 +1,209 @@
+
+        Returns:
+            Huggingface inference client.
+        """
+        return self.hf_endpoint.client
+
+    def provision(self) -> None:
+        """Provision or update remote Huggingface deployment instance.
+
+        This should then match the current configuration.
+        """
+
+        _ = create_inference_endpoint(
+            name=self.config.endpoint_name,
+            repository=self.config.repository,
+            framework=self.config.framework,
+            accelerator=self.config.accelerator,
+            instance_size=self.config.instance_size,
+            instance_type=self.config.instance_type,
+            region=self.config.region,
+            vendor=self.config.vendor,
+            account_id=self.config.account_id,
+            min_replica=self.config.min_replica,
+            max_replica=self.config.max_replica,
+            revision=self.config.revision,
+            task=self.config.task,
+            custom_image=self.config.custom_image,
+            type=self.config.endpoint_type,
+            namespace=self.config.namespace,
+            token=self.config.token,
+        ).wait(timeout=POLLING_TIMEOUT)
+
+        if self.hf_endpoint.url is not None:
+            logger.info(
+                "Huggingface inference endpoint successfully deployed."
+            )
+        else:
+            logger.info(
+                "Failed to start huggingface inference endpoint service..."
+            )
+
+    def check_status(self) -> Tuple[ServiceState, str]:
+        """Check the the current operational state of the HuggingFace deployment.
+
+        Returns:
+            The operational state of the HuggingFace deployment and a message
+            providing additional information about that state (e.g. a
+            description of the error, if one is encountered).
+        """
+        try:
+            _ = self.hf_endpoint.status
+        except (InferenceEndpointError, HfHubHTTPError):
+            return (ServiceState.INACTIVE, "")
+
+        if self.hf_endpoint.status == "running":
+            return (
+                ServiceState.ACTIVE,
+                f"HuggingFace Inference Endpoint deployment is available",
+            )
+
+        if self.hf_endpoint.status == "failed":
+            return (
+                ServiceState.ERROR,
+                f"HuggingFace Inference Endpoint deployment failed: ",
+            )
+
+        if self.hf_endpoint.status == "pending":
+            return (
+                ServiceState.PENDING_STARTUP,
+                "HuggingFace Inference Endpoint deployment is being created: ",
+            )
+
+    def deprovision(self, force: bool = False) -> None:
+        """Deprovision the remote HuggingFace deployment instance.
+
+        Args:
+            force: if True, the remote deployment instance will be
+                forcefully deprovisioned.
+        """
+        try:
+            self.hf_endpoint.delete()
+        except HfHubHTTPError:
+            pass
```

Comment:
We should log a `logger.error` or even raise an exception in this right? Explaining what happens and maybe internal details of the error

Reply:
Added it in.